src/python/grpcio_tests/tests/unit/_host_test.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2016 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests server and client side host API."""
+
+import unittest
+import weakref
+
+import grpc
+from grpc import _channel
+
+from tests.unit import test_common
+from tests.unit.framework.common import test_constants
+
+_HOSTNAME = 'testhost'
+
+_REQUEST = b'\x00\x00\x00'
+_RESPONSE = b'\x00\x00\x00'
+
+_UNARY_UNARY = '/test/UnaryUnary'
+_UNARY_STREAM = '/test/UnaryStream'
+_STREAM_UNARY = '/test/StreamUnary'
+_STREAM_STREAM = '/test/StreamStream'
+
+
+class _MethodHandler(grpc.RpcMethodHandler):
+
+    def __init__(self):
+        self.request_streaming = False
+        self.response_streaming = False
+        self.request_deserializer = None
+        self.response_serializer = None
+        self._host = None
+
+    def unary_unary(self, request, servicer_context):
+        self._host = servicer_context.host()
+        return _RESPONSE
+
+    def unary_stream(self, request, servicer_context):
+        self._host = servicer_context.host()
+        for _ in range(test_constants.STREAM_LENGTH):
+            yield _RESPONSE
+
+    def stream_unary(self, request_iterator, servicer_context):
+        self._host = servicer_context.host()
+        # TODO(issue:#6891) We should be able to remove this loop
+        for request in request_iterator:
+            pass
+        return _RESPONSE
+
+    def stream_stream(self, request_iterator, servicer_context):
+        self._host = servicer_context.host()
+        # TODO(issue:#6891) We should be able to remove this loop,
+        # and replace with return; yield
+        for request in request_iterator:
+            yield _RESPONSE
+
+    def host(self):
+        return self._host
+
+
+class _GenericHandler(grpc.GenericRpcHandler):
+
+    def __init__(self):
+        self._handler = _MethodHandler()
```

Comment:
`self._host_cell = [None]`? Then in `service` you can just have `return _MethodHandler(self._host_cell, <False or True>, <False or True>)`?